### DISCUSSION TITLE
Simplify generate.sh

### DISCUSSION THREAD

**Discussion by ggwpez** [hyperledger-labs/go-perun#46] [General]
Location: `generate.sh`

Description:
Use the following snippet to get rid of the `awk` call and simplify the script.

Snippet from @sebastianst, [link](https://git.perun.network/-/snippets/52):  
```sh
#!/bin/bash

# SPDX-License-Identifier: Apache-2.0

set -e

# Configuration
solpath="../contracts/contracts"
bindpath="../../bindings" # reverse to solpath
solc="solc-static-linux"

# Download solc.
wget -nc "https://github.com/ethereum/solidity/releases/download/v0.8.3/${solc}"
chmod +x $solc
echo -e "⚠ Ensure that the newest version of abigen is installed"

# Compile in solpath
cd "${solpath}"
"${bindpath}/${solc}" --combined-json abi,bin,bin-runtime \
  --allow-paths '.,../node_modules/@openzeppelin' \
  --optimize --optimize-runs 10000 \
  --overwrite -o "${bindpath}" \
  "@openzeppelin=../node_modules/@openzeppelin" \
  *.sol

# Bind
cd "${bindpath}"
abigen --pkg bindings --out bindings.go --combined-json combined.json \
  --exc Address,Bytes,Context,ECDSA,ERC165,ERC20,ERC721,IERC165,IERC721Enumerable,IERC721Metadata,IERC721Receiver,Ownable,Sig,Strings

echo -e "🎉 Bindings generated"
```

**Comment by sebastianst** [comment]
Did you try it out yet @ggwpez ? I think at least the `--exc` argument needs adjustment. Although they seem to be ignored right now anyways...

**Comment by ggwpez** [comment]
> Did you try it out yet @ggwpez ? I think at least the `--exc` argument needs adjustment. Although they seem to be ignored right now anyways...

I did not try it yet, just wanted to put have it written down as an issue. @sebastianst

**Comment by sebastianst** [comment]
BTW fixed the behavior of `--exc` in https://github.com/ethereum/go-ethereum/pull/22620

**Comment by matthiasgeihs** [comment]
We are currently not using `awk` in script `generate.sh`.

To be honest, I don't think our script is very complicated. The most complicated part is generating the BinRuntime.go file for each contract, but I don't think this is solved in the snippet, right?

**Comment by sebastianst** [comment]
The current `generate.sh` script has two related issues. The [`abigen call`](https://github.com/hyperledger-labs/go-perun/blob/9923f1108ac55b08b52d68dbc29ea4d23ead5902/backend/ethereum/bindings/generate.sh#L48)
```sh
    $ABIGEN --pkg $PKG --sol ../contracts/contracts/$FILE.sol --out $PKG/$FILE.go --solc $SOLC
```
passes the solc compiler via `--solc`. Unfortunately, this method doesn't allow setting compiler flags. So in particular, no `--optimize-runs 10000` can be specified. This means that contracts deployed using the bindings are not optimized, which is currently not an issue, but when we move towards production, the contracts should be optimized to minimize transaction costs.

The second issue is that the generation of the `BinRuntime.go`s use a second direct `solc` call, possibly with different options that `abigen` internally. So the resulting deployed-bytecode checks may fail. It was unclean in the first place to compile the contracts _twice_, once for bindings generation and a second time for binruntime generation. Using `--combined-json` is therefore cleaner and more consistent (and faster, because the whole contracts tree is only compiled once). Only thing that would be left to do is extract the bin-runtime from the `combined.json`, which should be possible with a little [`jq`](https://stedolan.github.io/jq/) magic :sparkles: 

It's not an urgent issue to fix, but I reopen it as to not forget about it.

**Reply by matthiasgeihs** [reply]
OK. So there are two issues:

1. We run solc twice (once with abigen for the bindings, once for the binary runtime), potentially creating inconsistencies.
2. solc with abigen does not allow configuration parameters

I will create a github issue for these and open a PR.

**Comment by ggwpez** [comment]
>Using --combined-json is therefore cleaner and more consistent (and faster, because the whole contracts tree is only compiled once)

This would also help to verify the Code on Etherscan, right? @seb  
I think @matthiasgeihs tried it by combining the contracts but it is needlessly difficult.  

PS: Feel free to edit the issue.

**Reply by sebastianst** [reply]
No it wouldn't help. It was a misunderstanding from my side that you can use the `combined.json` for etherscan. They need an "input json" with the source code and compiler flags, not the "output" that this `combined.json` is.
